NRE in generator

I saw this here: https://github.com/xamarin/AndroidX/pull/247

```
    BindingsGenerator
        ...
        Warnings
            /Users/runner/work/1/s/source/Metadata.common.xml(2, 6) warning BG8A00: Metadata.xml element '<remove-node path="/api/package/*[contains(@name, 'IcsImpl') or contains(@name, 'JellyBeanImpl') or contains(@name, 'JellyBeanMr2Impl') or contains(@name, 'Api16') or contains(@name, 'Api17') or contains(@name, 'Api18') or contains(@name, 'Api19') or contains(@name, 'Api20') or contains(@name, 'Api21') or contains(@name, 'Api22') or contains(@name, 'Api23') or contains(@name, 'Api24') or contains(@name, 'Api25') or contains(@name, 'Api26') or contains(@name, 'Api27') or contains(@name, 'Api28') or contains(@name, 'Api29') or contains(@name, 'Api30') or contains(@name, 'Honeycomb') or contains(@name, 'IceCreamSandwich') or contains(@name, 'Jellybean') or contains(@name, 'KitKat') or contains(@name, 'Lollipop') or contains(@name, 'Marshmallow') or contains(@name, 'Nougat') or contains(@name, 'Oreo') or contains(@name, 'ImplJB') or contains(@name, 'ImplICS') or contains(@name, 'ImplIcs') or contains(@name, 'ImplHC') or contains(@name, 'HcMr1') or contains(@name, 'HcMr2') or contains(@name, 'HcMr3') or contains(@name, 'JbMr2') or contains(@name, 'JbMr2') or contains(@name, 'JbMr1') or contains(@name, 'ICSMr1') or contains(@name, 'ICSMr2') or contains(@name, 'ImplV11') or contains(@name, 'ImplV21')]" />' matched no nodes. [/Users/runner/work/1/s/generated/androidx.tvprovider.tvprovider/androidx.tvprovider.tvprovider.csproj]
            /Users/runner/work/1/s/source/Metadata.common.xml(2, 6): warning BG8A00: Metadata.xml element '<remove-node path="/api/package/*[contains(@name, 'IcsImpl') or contains(@name, 'JellyBeanImpl') or contains(@name, 'JellyBeanMr2Impl') or contains(@name, 'Api16') or contains(@name, 'Api17') or contains(@name, 'Api18') or contains(@name, 'Api19') or contains(@name, 'Api20') or contains(@name, 'Api21') or contains(@name, 'Api22') or contains(@name, 'Api23') or contains(@name, 'Api24') or contains(@name, 'Api25') or contains(@name, 'Api26') or contains(@name, 'Api27') or contains(@name, 'Api28') or contains(@name, 'Api29') or contains(@name, 'Api30') or contains(@name, 'Honeycomb') or contains(@name, 'IceCreamSandwich') or contains(@name, 'Jellybean') or contains(@name, 'KitKat') or contains(@name, 'Lollipop') or contains(@name, 'Marshmallow') or contains(@name, 'Nougat') or contains(@name, 'Oreo') or contains(@name, 'ImplJB') or contains(@name, 'ImplICS') or contains(@name, 'ImplIcs') or contains(@name, 'ImplHC') or contains(@name, 'HcMr1') or contains(@name, 'HcMr2') or contains(@name, 'HcMr3') or contains(@name, 'JbMr2') or contains(@name, 'JbMr2') or contains(@name, 'JbMr1') or contains(@name, 'ICSMr1') or contains(@name, 'ICSMr2') or contains(@name, 'ImplV11') or contains(@name, 'ImplV21')]" />' matched no nodes.
        Unhandled Exception:
        Unhandled Exception:
        System.NullReferenceException: Object reference not set to an instance of an object
        System.NullReferenceException: Object reference not set to an instance of an object
        at generator.SourceWriters.BoundInterface.AddInheritedInterfaces (MonoDroid.Generation.InterfaceGen iface, MonoDroid.Generation.CodeGenerationOptions opt) [0x00034] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundInterface.AddInheritedInterfaces (MonoDroid.Generation.InterfaceGen iface, MonoDroid.Generation.CodeGenerationOptions opt) [0x00034] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundInterface..ctor (MonoDroid.Generation.InterfaceGen iface, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x0019b] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundInterface..ctor (MonoDroid.Generation.InterfaceGen iface, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x0019b] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.SourceWriterExtensions.BuildManagedTypeModel (MonoDroid.Generation.GenBase gen, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x0001e] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.SourceWriterExtensions.BuildManagedTypeModel (MonoDroid.Generation.GenBase gen, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x0001e] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundClass.AddNestedTypes (MonoDroid.Generation.ClassGen klass, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x00042] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundClass.AddNestedTypes (MonoDroid.Generation.ClassGen klass, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x00042] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundClass..ctor (MonoDroid.Generation.ClassGen klass, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo generationInfo) [0x00338] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundClass..ctor (MonoDroid.Generation.ClassGen klass, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo generationInfo) [0x00338] in <1cf370e892674e6d8b798a230e594994>:0 
        at MonoDroid.Generation.JavaInteropCodeGenerator.WriteType (MonoDroid.Generation.GenBase gen, System.String indent, MonoDroid.Generation.GenerationInfo gen_info) [0x00037] in <1cf370e892674e6d8b798a230e594994>:0 
        at MonoDroid.Generation.JavaInteropCodeGenerator.WriteType (MonoDroid.Generation.GenBase gen, System.String indent, MonoDroid.Generation.GenerationInfo gen_info) [0x00037] in <1cf370e892674e6d8b798a230e594994>:0 
        at MonoDroid.Generation.ClassGen.Generate (MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.GenerationInfo gen_info) [0x0007d] in <1cf370e892674e6d8b798a230e594994>:0 
        at MonoDroid.Generation.ClassGen.Generate (MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.GenerationInfo gen_info) [0x0007d] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Run (Xamarin.Android.Binder.CodeGeneratorOptions options, Java.Interop.Tools.Cecil.DirectoryAssemblyResolver resolver) [0x005d1] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Run (Xamarin.Android.Binder.CodeGeneratorOptions options, Java.Interop.Tools.Cecil.DirectoryAssemblyResolver resolver) [0x005d1] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Run (Xamarin.Android.Binder.CodeGeneratorOptions options) [0x0001b] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Run (Xamarin.Android.Binder.CodeGeneratorOptions options) [0x0001b] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Main (System.String[] args) [0x0000c] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Main (System.String[] args) [0x0000c] in <1cf370e892674e6d8b798a230e594994>:0 
        [ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
        [ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
        at generator.SourceWriters.BoundInterface.AddInheritedInterfaces (MonoDroid.Generation.InterfaceGen iface, MonoDroid.Generation.CodeGenerationOptions opt) [0x00034] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundInterface.AddInheritedInterfaces (MonoDroid.Generation.InterfaceGen iface, MonoDroid.Generation.CodeGenerationOptions opt) [0x00034] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundInterface..ctor (MonoDroid.Generation.InterfaceGen iface, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x0019b] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundInterface..ctor (MonoDroid.Generation.InterfaceGen iface, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x0019b] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.SourceWriterExtensions.BuildManagedTypeModel (MonoDroid.Generation.GenBase gen, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x0001e] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.SourceWriterExtensions.BuildManagedTypeModel (MonoDroid.Generation.GenBase gen, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x0001e] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundClass.AddNestedTypes (MonoDroid.Generation.ClassGen klass, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x00042] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundClass.AddNestedTypes (MonoDroid.Generation.ClassGen klass, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo genInfo) [0x00042] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundClass..ctor (MonoDroid.Generation.ClassGen klass, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo generationInfo) [0x00338] in <1cf370e892674e6d8b798a230e594994>:0 
        at generator.SourceWriters.BoundClass..ctor (MonoDroid.Generation.ClassGen klass, MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.CodeGeneratorContext context, MonoDroid.Generation.GenerationInfo generationInfo) [0x00338] in <1cf370e892674e6d8b798a230e594994>:0 
        at MonoDroid.Generation.JavaInteropCodeGenerator.WriteType (MonoDroid.Generation.GenBase gen, System.String indent, MonoDroid.Generation.GenerationInfo gen_info) [0x00037] in <1cf370e892674e6d8b798a230e594994>:0 
        at MonoDroid.Generation.JavaInteropCodeGenerator.WriteType (MonoDroid.Generation.GenBase gen, System.String indent, MonoDroid.Generation.GenerationInfo gen_info) [0x00037] in <1cf370e892674e6d8b798a230e594994>:0 
        at MonoDroid.Generation.ClassGen.Generate (MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.GenerationInfo gen_info) [0x0007d] in <1cf370e892674e6d8b798a230e594994>:0 
        at MonoDroid.Generation.ClassGen.Generate (MonoDroid.Generation.CodeGenerationOptions opt, MonoDroid.Generation.GenerationInfo gen_info) [0x0007d] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Run (Xamarin.Android.Binder.CodeGeneratorOptions options, Java.Interop.Tools.Cecil.DirectoryAssemblyResolver resolver) [0x005d1] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Run (Xamarin.Android.Binder.CodeGeneratorOptions options, Java.Interop.Tools.Cecil.DirectoryAssemblyResolver resolver) [0x005d1] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Run (Xamarin.Android.Binder.CodeGeneratorOptions options) [0x0001b] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Run (Xamarin.Android.Binder.CodeGeneratorOptions options) [0x0001b] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Main (System.String[] args) [0x0000c] in <1cf370e892674e6d8b798a230e594994>:0 
        at Xamarin.Android.Binder.CodeGenerator.Main (System.String[] args) [0x0000c] in <1cf370e892674e6d8b798a230e594994>:0 
        Errors
            /Library/Frameworks/Xamarin.Android.framework/Libraries/xbuild/Xamarin/Android/Xamarin.Android.Bindings.Core.targets(53,5): error MSB6006: "mono" exited with code 1. [/Users/runner/work/1/s/generated/androidx.tvprovider.tvprovider/androidx.tvprovider.tvprovider.csproj]
```

Using: https://dl.internalx.com/vsts-devdiv/Xamarin.Android/public/4451481/master/05bb8e0eae11ae6a73838b13cf91ee2433169dff/xamarin.android-11.2.99.85.pkg

Logs are probably easiest downloaded from Azure DevOps: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4456235&view=artifacts&pathAsName=false&type=publishedArtifacts

They are too big to upload here.

--- a/tests/generator-Tests/Unit-Tests/CodeGeneratorTests.cs
+++ b/tests/generator-Tests/Unit-Tests/CodeGeneratorTests.cs
@@ -229,6 +229,44 @@ namespace generatortests
 
 			StringAssert.DoesNotContain ("[global::System.Runtime.Versioning.SupportedOSPlatformAttribute (\"android30.0\")]", builder.ToString (), "Should contain SupportedOSPlatform!");
 		}
+
+		[Test]
+		public void InheritedInterfaceAsClass ()
+		{
+			// This is a somewhat cheating way to repro a real issue.
+			// The real issue is:
+			// - Binding an interface which implements 'android.provider.BaseColumns'
+			// - android.provider.BaseColumns is an interface in android.jar
+			// - Mono.Android.dll has both:
+			//   - [Register ("android.provider.BaseColumns")] public abstract class BaseColumns
+			//   - [Register ("android.provider.BaseColumns")] public interface IBaseColumns
+			// Our Java type resolution is "last one wins" and happens to pick the class instead of
+			// the interface. So our code is trying to bind an interface that implements a class.
+			var xml = @"<api>
+			  <package name='java.lang' jni-name='java/lang'>
+			    <class abstract='false' deprecated='not deprecated' final='false' name='Object' static='false' visibility='public' jni-signature='Ljava/lang/Object;' />
+			  </package>
+			  <package name='com.xamarin.android' jni-name='com/xamarin/android'>
+			    <class abstract='false' deprecated='not deprecated' extends='java.lang.Object' extends-generic-aware='java.lang.Object' final='false' name='MyConsts' static='false' visibility='public' jni-signature='Ljava/lang/Object;' />
+			    <interface abstract='false' deprecated='not deprecated' extends='java.lang.Object' extends-generic-aware='java.lang.Object' jni-extends='Ljava/lang/Object;' final='false' name='MyClass' static='false' visibility='public' jni-signature='Lcom/xamarin/android/MyClass;'>
+			      <implements name='com.xamarin.android.MyConsts' name-generic-aware='com.xamarin.android.MyConsts' jni-type='Lcom/xamarin/android/MyConsts;'></implements>
+			    </interface>
+			  </package>
+			</api>";
+
+			var gens = ParseApiDefinition (xml);
+
+			// Disable "BuildingCoreAssembly"
+			(gens.Single (g => g.Name == "Object") as ClassGen).FromXml = false;
+
+			var klass = gens.Single (g => g.Name == "IMyClass");
+
+			generator.Context.ContextTypes.Push (klass);
+			generator.WriteType (klass, string.Empty, new GenerationInfo ("", "", "MyAssembly"));
+			generator.Context.ContextTypes.Pop ();
+
+			Assert.Pass ("WriteType did not NRE");
+		}
 	}
 
 	[TestFixture]

--- a/tools/generator/SourceWriters/BoundInterface.cs
+++ b/tools/generator/SourceWriters/BoundInterface.cs
@@ -137,7 +137,11 @@ namespace generator.SourceWriters
 			foreach (var isym in iface.Interfaces) {
 				var igen = (isym is GenericSymbol ? (isym as GenericSymbol).Gen : isym) as InterfaceGen;
 
-				if (igen.IsConstSugar (opt) || igen.RawVisibility != "public")
+				// igen *should not* be null here because we *should* only be inheriting interfaces, but
+				// in the case of constants on that interface we create a C# *class* that is registered for the
+				// Java *interface*. Thus when we do type resolution, we are actually pointed to the
+				// Foo abstract class instead of the IFoo interface.
+				if (igen is null || igen.IsConstSugar (opt) || igen.RawVisibility != "public")
 					continue;
 
 				Implements.Add (opt.GetOutputName (isym.FullName));

